File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.js
function solution(id_list, report, k) {
    // 1. 중복 제거
    const uniqueReports = [...new Set(report)];

    // 2. 신고 관계 저장
    const reportedBy = {};  // {신고당한 사람: [신고한 사람들]}
    const reportedCount = {}; // {유저: 신고당한 횟수}

    id_list.forEach(id => {
        reportedBy[id] = [];
        reportedCount[id] = 0;
    });

    uniqueReports.forEach(r => {
        const [user, target] = r.split(" ");
        reportedBy[target].push(user);
        reportedCount[target]++;
    });

    // 3. 정지된 유저 찾기
    const banned = id_list.filter(user => reportedCount[user] >= k);

    // 4. 메일 횟수 계산
    const idIndex = {};
    id_list.forEach((id, i) => idIndex[id] = i);
    const answer = new Array(id_list.length).fill(0);

    banned.forEach(b => {
        reportedBy[b].forEach(reporter => {
            answer[idIndex[reporter]]++;
        });
    });

    return answer;
}

File: 프로그래머스/1/92334. 신고 결과 받기/신고 결과 받기.kt
class Solution {
    fun solution(id_list: Array<String>, report: Array<String>, k: Int): IntArray {
        val answer = IntArray(id_list.size)

        // 1. 중복 신고 제거
        val uniqueReports = report.toSet()

        // 2. 신고 관계 저장 (신고당한 사람 -> 신고한 사람들 집합)
        val reportedBy = mutableMapOf<String, MutableSet<String>>()
        id_list.forEach { id ->
            reportedBy[id] = mutableSetOf()
        }

        uniqueReports.forEach { r ->
            val (reporter, target) = r.split(" ")
            reportedBy[target]?.add(reporter)
        }

        // 3. 정지된 유저 찾기
        val banned = reportedBy.filter { it.value.size >= k }.keys

        // 4. 메일 횟수 계산
        val idIndex = id_list.withIndex().associate { it.value to it.index }
        banned.forEach { user ->
            reportedBy[user]?.forEach { reporter ->
                val idx = idIndex[reporter]!!
                answer[idx]++
            }
        }

        return answer
    }
}
